### PR TITLE
OCPQE-26389: aws - collect IPI cluster launcher info to help cost usage analysis

### PR DESCRIPTION
**Pull request by jianlinliu**
Use userTags to tag resourced created by IPI to help analyze cost usage.

For job launched by pj-rehearse, the added userTags is:
```
usage-cluster-type classic
usage-job-type presubmit
usage-user ${pr-author}
usage-ci-type prow
usage-pull-request ${pr-number}
```
For job launched by periodic ci, the added userTags is:
```
usage-cluster-type classic
usage-job-type periodic
usage-user ${prow-repo}-${pro-repo-branch}
usage-ci-type prow
```

### CODE REVIEW

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-aws-ipi-amd-f28-netobserv

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-aws-ipi-amd-f28-netobserv

**Comment by jianlinliu** [comment]
cc @yunjiang29 for early review, rehearse job log is [here](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/57922/rehearse-57922-periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-aws-ipi-amd-f28-netobserv/1846868388438085632/artifacts/aws-ipi-amd-f28-netobserv/ipi-install-install/artifacts/install-config.yaml).

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    user="cron"
+fi
+ci_type="prow"
+if [[ "${JOB_NAME_SAFE:-}" == "launch" ]]; then
+    ci_type="cluster-bot"
+fi
+
+cluster_info_file="${ARTIFACT_DIR}/aws-cluster-info-user-tag"
+cat <<EOF >>"${cluster_info_file}"
+cluster-type ${CI_CLUSTER_TYPE}
```
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:35]: add a prefix for these keys? like `info-` or something like that?

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+EOF
+
+if (( ocp_minor_version > 12 || ocp_major_version > 4 )); then
+    while read -r TAG VALUE
+    do
+      printf 'Setting user tag to help cost usage analysis - %s: %s\n' "${TAG}" "${VALUE}"
+      yq-go write -i "${CONFIG}" "platform.aws.userTags.${TAG}" "${VALUE}"
+    done < "${cluster_info_file}"
+else
+    echo "userTags on aws get supported from 4.12, skip..."
```
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:49]: I think we can put this before `cluster_info_file="${ARTIFACT_DIR}/aws-cluster-info-user-tag"`, so that `aws-cluster-info-user-tag` wion't be generated

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-ref.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    name: "4.16"
+    tag: upi-installer
+  commands: ipi-conf-aws-launcher-info-commands.sh
+  resources:
+    requests:
+      cpu: 10m
+      memory: 100Mi
+  env:
+  - name: CI_CLUSTER_TYPE
+    default: "classic"
```
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-ref.yaml:14]: we can add a description to indicate that 5 tags will be added to the cluster.

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    ci_type="cluster-bot"
+fi
+
+cluster_info_file="${ARTIFACT_DIR}/aws-cluster-info-user-tag"
+cat <<EOF >>"${cluster_info_file}"
+cluster-type ${CI_CLUSTER_TYPE}
+job-type ${job_type}
+user ${user}
+pull-request ${pull_number}
+ci ${ci_type}
```
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:39]: Make key name self-descriptive？
```suggestion
ci-type ${ci_type}
```

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 2 above)
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:49]: the limit of tags was extended to 25 in 4.11.z [1], but it's OK as 4.11 is EOL now.


[1] https://github.com/openshift/installer/pull/6303

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+EOF
+
+if (( ocp_minor_version > 12 || ocp_major_version > 4 )); then
```
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:42]:
```suggestion
if (( ocp_minor_version > 10 || ocp_major_version > 4 )); then
```
or
```suggestion
if (( ocp_minor_version > 11 || ocp_major_version > 4 )); then
```

**Review by yunjiang29 — COMMENTED** [review]
@jianlinliu basically, the PR looks good to me, I added some minor comments below, PTAL

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 1 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:35]: how about `usage-`

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 5 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:42]: aws user tag get supported from 4.11 ?

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 2 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:49]: Actually I saved `cluster_info_file` on purpose even some OCP version does not support it, if needed, we can easily figure out who trigger the jobs when we go through the step artifacts dir.

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ocp_major_version=$( echo "${ocp_version}" | awk --field-separator=. '{print $1}' )
+ocp_minor_version=$( echo "${ocp_version}" | awk --field-separator=. '{print $2}' )
+rm /tmp/pull-secret
+
+job_type="${JOB_TYPE:-}"
+user=""
+pull_number=""
+if [[ "${job_type}" == "presubmit" ]]; then
+    user=$(echo "${JOB_SPEC:-}" | jq -r '.refs.pulls[].author')
+    pull_number=${PULL_NUMBER:-unknow}
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:23]: typo for `unknow`....should be `unknown`

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+fi
+if [[ "${job_type}" == "periodic" ]]; then
+    user="cron"
+fi
+ci_type="prow"
+if [[ "${JOB_NAME_SAFE:-}" == "launch" ]]; then
+    ci_type="cluster-bot"
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:30]: should be use `pr-cluster-bot` instead, to differentiate the normal `cluster-bot`?

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:30]: what means `pr-cluster-bot` ?

Comment (pruan-rht) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:30]: @jianlinliu I'm assuming that `launch` is invoked by the Pull Request, thus the `pr-` prefix.   Where does `JOB_NAME_SAFE` come from?

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh:30]: No `launch` means created by robot, `JOB_NAME_SAFE` is from prow ci ENV. For the job launched by pull request, they are indicated by `job_type` as `presubmit`.

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 1 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: Done

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 4 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: Done

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 5 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: Done

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 6 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: Done

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 7 above)
Comment (pruan-rht) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: ok, then leaving it as it's works for me...

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+EOF
+
+if (( ocp_minor_version > 10 || ocp_major_version > 4 )); then
+    while read -r TAG VALUE
+    do
+      printf 'Setting user tag to help cost usage analysis - %s: %s\n' "${TAG}" "${VALUE}"
+      yq-go write -i "${CONFIG}" "platform.aws.userTags.${TAG}" "${VALUE}"
+    done < "${cluster_info_file}"
+else
+    echo "userTags on aws get supported from 4.12, skip..."
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:51]: perhaps reword it to 
```
USER_TAGS is only supported for OCP version >= 4.12, skipping...
```

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh]: Done.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.metadata.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.metadata.json:1]: @jianlinliu I think for JSON files, the indentation is 4 spaces, not 8 as what's there currently.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+ref:
+  as: ipi-conf-aws-usage-info
+  from_image:
+    namespace: ocp
+    name: "4.16"
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml:5]: should this be hard-coded as `4.16`?

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  commands: ipi-conf-aws-usage-info-commands.sh
+  resources:
+    requests:
+      cpu: 10m
+      memory: 100Mi
+  env:
+  - name: USAGE_CLUSTER_TYPE
+    default: "classic"
+    documentation: |-
+      Indicate this cluster is a classic cluster, or hypershift-mgmt cluster
```
Comment (pruan-rht) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml:16]: should we list all the supported cluster_type here?

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml (hunk 10 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml:5]: Yeah, the step does not care about the version of image.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.metadata.json (hunk 9 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.metadata.json:1]: The file is automatically generated by prow make command.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml (hunk 11 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-ref.yaml:16]: No, not each type of cluster is supposed to use this step, roas, hypershift-hosted cluster is not supposed to use this step, because they are not using userTags in install-config.yaml.

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-multi-nightly-aws-ipi-amd-f28-netobserv

File: ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh (hunk 2 above)
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/conf/aws/launcher-info/ipi-conf-aws-launcher-info-commands.sh]: Got it.

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-stable-aws-ipi-usertags-custom-sg-f14

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+rm /tmp/pull-secret
+
+job_type="${JOB_TYPE:-}"
+user=""
+if [[ "${job_type}" == "presubmit" ]]; then
+    author=$(echo "${JOB_SPEC:-}" | jq -r '.refs.pulls[].author')
+    pull_number=${PULL_NUMBER:-unknown}
+    user="${author}-${pull_number}"
+fi
+if [[ "${job_type}" == "periodic" ]]; then
```
Comment (aleskandro) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:27]: Should we consider postsubmit too?

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:27]: when postsubmit job would be triggered ?

Comment (aleskandro) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:27]: Yes but also an else to catch any possible job type different than the ones considered?

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:27]: Do you have an example postsubmit job link, so that I can figure out what tag can be filled there.

Comment (aleskandro) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/branch-ci-openshift-assisted-service-release-ocm-2.10-mce-publish/1848258653547663360

Comment (aleskandro) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh]: (I guess) they should be all the ones with `branch-ci` at the beginning.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+job_type="${JOB_TYPE:-}"
+user=""
+pull_number="null"
+if [[ "${job_type}" == "presubmit" ]]; then
+    user=$(echo "${JOB_SPEC:-}" | jq -r '.refs.pulls[].author')
+    pull_number=${PULL_NUMBER:-unknown}
+elif [[ "${job_type}" == "periodic" ]]; then
+    user=$(echo "${JOB_SPEC:-}" | jq -r '.extra_refs[]' |  jq -r '.repo + "-" + .base_ref')
+else
+    user="null"
```
Comment (aleskandro) [inline ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh:29]:
```suggestion
    user=""
```

better an empty string on the pandas side. user will always be a string column.

the installer should not complain about the empty value.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh]: done.

File: ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh (hunk 12 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/usage-info/ipi-conf-aws-usage-info-commands.sh]: Done.

**Review by aleskandro — COMMENTED** [review]
/lgtm

Thanks @jianlinliu

**Comment by jianlinliu** [comment]
/pj-rehearse pull-ci-openshift-jaegertracing-jaeger-operator-main-ocp-4.13-hypershift-jaeger-e2e-tests

**Comment by jianlinliu** [comment]
@heliubj18 @LiangquanLi930 can you help check the change for hypershift parts, I only added `hypershift-mgmt` cluster type for management cluster, if I am wrong or miss something, pls correct me.

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-arm64-stable-aws-ipi-usertags-custom-sg-f14

**Comment by heliubj18** [comment]
LGTM

**Comment by jianlinliu** [comment]
@yunjiang29 do you mind to review again, I injected `ipi-conf-aws-usage-info` to more install workflow, it is ready to merge.

/pj-rehearse ack

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/usertags/custom-security-groups/provision/cucushift-installer-rehearse-aws-ipi-usertags-custom-security-groups-provision-chain.yaml [hunk 14]
```diff
@@ -37,14 +38,9 @@ chain:
...
       key853 val853
-      key624 val624
-      key645 val645
-      key100 val100
-      key187 val187
-      key250 val250
       key155 val155
       key863 val863
-    documentation: "UserTags for cluster, 23 user tags + 1 expirationDate tag + 1 clusterName tag = 25 totally"
+    documentation: "UserTags for cluster, 18 user tags + 1 expirationDate tag + 1 clusterName tag = 20 totally, leave 5 userTags quota for ipi-conf-aws-usage-info"
```
Comment (yunjiang29) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/usertags/custom-security-groups/provision/cucushift-installer-rehearse-aws-ipi-usertags-custom-security-groups-provision-chain.yaml:43]: 👍
